File: pubsub_py/pubsub_py/sub_diy_msg.py
#! /usr/bin/env python3
# -*- coding: utf-8 -*-

import rclpy
from rclpy.node import Node

from diy_interface.msg import Student
from diy_interface.msg import Sphere

class Subscriber(Node):
    def __init__(self):
        super().__init__('test_sub_diy_msg')
        self._sub_student = self.create_subscription(Student, "student_topic", self.student_topic_callback, 10)
        self._sub_sphere = self.create_subscription(Sphere, "sphere_topic", self.sphere_topic_callback, 10)

    def student_topic_callback(self, msg):
        self.get_logger().info("i received student in python: {0}, {1}".format(msg.name, msg.age))

    def sphere_topic_callback(self, msg):
        self.get_logger().info("i received sphere in python: ({0}, {1}, {2}), {3}".format(msg.center.x, msg.center.y, msg.center.z, msg.radius))


def main(args=None):
    rclpy.init(args=args)
    
    sub_node = Subscriber()
    rclpy.spin(sub_node)

    sub_node.destroy_node()
    rclpy.shutdown()


if __name__ == '__main__':
    main()

File: pubsub_cpp/src/sub_diy_msg.cpp
#include <memory>

#include "rclcpp/rclcpp.hpp"
#include "diy_interface/msg/student.hpp"
#include "diy_interface/msg/sphere.hpp"

class Subscriber : public rclcpp::Node 
{
public:
    Subscriber() : Node("test_sub_diy_msg") {
        sub_student_ = this->create_subscription<diy_interface::msg::Student>(
            "student_topic", 10, std::bind(&Subscriber::student_topic_callback, this, std::placeholders::_1));
        sub_sphere_ = this->create_subscription<diy_interface::msg::Sphere>(
            "sphere_topic", 10, std::bind(&Subscriber::sphere_topic_callback, this, std::placeholders::_1));
    }

private:
    void student_topic_callback(const diy_interface::msg::Student &msg) const {
        RCLCPP_INFO(this->get_logger(), "i received student in cpp: %s, %d", msg.name.c_str(), msg.age);
    }
    void sphere_topic_callback(const diy_interface::msg::Sphere &msg) const {
        RCLCPP_INFO(this->get_logger(), "i received sphere in cpp: (%f, %f, %f), %f", 
            msg.center.x, msg.center.y, msg.center.z, msg.radius);
    }

private:
    rclcpp::Subscription<diy_interface::msg::Student>::SharedPtr sub_student_;
    rclcpp::Subscription<diy_interface::msg::Sphere>::SharedPtr sub_sphere_;
};


int main(int argc, char * argv[]) {
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<Subscriber>());
    rclcpp::shutdown();

    return 0;
}